activerecord/lib/active_record/connection_handling.rb

```diff
@@ -87,7 +87,36 @@ def config
     # also be used to "borrow" the connection to do database work unrelated
     # to any of the specific Active Records.
     def connection
-      retrieve_connection
+      connection_mutex.synchronize do
+        @temp_conn_spec_name ||= nil
+        retrieve_connection(@temp_conn_spec_name)
+      end
+    end
+
+    def using_connection(spec_name) # :nodoc:
+      connection_mutex.synchronize do
```

Comment:
Using a mutex feels wrong here. This is a thread local value, we shouldn't need to block other threads to change it.